### PR TITLE
Updated GuiTabs PR to 1.11.x

### PR DESCRIPTION
**Pull request by AshIndigo**
Changes made from #3349:
- Modified code to work with 1.11 changes
- Added back continuities (You can go back to the vanilla tab easily without exiting the GUI)
- Improved and added Java docs to various methods and fields.
- Made a test mod which will check if the tabs are working and if ResourceLocation and ItemStack icons are working too
 
Credits to @ArtixAllMighty for making the original PR

### CODE REVIEW

File: src/test/java/net/minecraftforge/test/GuiTabsTest.java [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+                    @Override
+                    public void onTabClicked()
+                    {
+                        GuiTabsTest.INSTANCE.sendToServer(new TabMessageHandler.TabMessage(1));
+                    }
+                }.setTargetGui(GuiChest.class).addTo(GuiInventory.class);
+            }
+            for (int i = 0; 13 != i; i++)
+            {
+                new GuiTab("Tab3", new ItemStack(Blocks.FURNACE), TestGui2.class) {
```
Comment (HenryLoenwind) [inline src/test/java/net/minecraftforge/test/GuiTabsTest.java]: Have you tested items with enchantment glint?

File: src/main/java/net/minecraftforge/client/player/inventory/tabs/GuiTab.java [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+        this.iconStack = icon;
+        // Make sure the vanilla tab doesnt get duplicated
+        if (GuiTab.tabRegistry.get(clazz).contains(VANILLA_INVENTORY_TAB) == false)
+        {
+            GuiTab.VANILLA_INVENTORY_TAB.addTo(clazz);
+        }
+    }
+
+    /**
+     * ItemStack version of tabs
```
Comment (Alpvax) [inline src/main/java/net/minecraftforge/client/player/inventory/tabs/GuiTab.java]: Should say ResourceLocation version

File: src/test/java/net/minecraftforge/test/GuiTabsTest.java (hunk 1 above)
Comment (AshIndigo) [reply src/test/java/net/minecraftforge/test/GuiTabsTest.java]: I added a test to one of the items and checked in-game. The glows do work

File: src/main/java/net/minecraftforge/client/player/inventory/tabs/GuiTab.java (hunk 2 above)
Comment (AshIndigo) [reply src/main/java/net/minecraftforge/client/player/inventory/tabs/GuiTab.java]: Fixed

**Comment by randysdoom** [comment]
Any idea as to when/if this will be pulled?

**Comment by liach** [comment]
~~@mezz If this is okay, please assign Lex~~
Not really... There are some issues.

File: patches/minecraft/net/minecraft/client/gui/inventory/GuiInventory.java.patch [hunk 3]
```diff
@@ -0,0 +1,13 @@
+--- ../src-base/minecraft/net/minecraft/client/gui/inventory/GuiInventory.java
++++ ../src-work/minecraft/net/minecraft/client/gui/inventory/GuiInventory.java
+@@ -112,8 +112,10 @@
+         GlStateManager.func_179138_g(OpenGlHelper.field_77478_a);
+     }
+ 
++    @Override
```
Comment (liach) [inline patches/minecraft/net/minecraft/client/gui/inventory/GuiInventory.java.patch]: Don't add this

File: src/main/java/net/minecraftforge/client/player/inventory/tabs/GuiTab.java [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+import net.minecraft.item.ItemStack;
+import net.minecraft.util.ResourceLocation;
+import net.minecraftforge.fml.common.FMLLog;
+import net.minecraftforge.fml.relauncher.Side;
+import net.minecraftforge.fml.relauncher.SideOnly;
+
+/**
+ * Class used to add a new GuiTab
+ * 
+ * @author Subaraki, Modified by Ash Indigo
```
Comment (liach) [inline src/main/java/net/minecraftforge/client/player/inventory/tabs/GuiTab.java]: `@author` tags?

File: src/main/java/net/minecraftforge/client/player/inventory/tabs/GuiTab.java [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+     *            The icon of the tab
+     * @param clazz
+     *            The class that is going to be displayed
+     */
+    public GuiTab(String name, ResourceLocation icon, Class<? extends GuiContainer> clazz)
+    {
+        this.name = name;
+        this.iconResLoc = icon;
+        // Make sure the vanilla tab doesnt get duplicated
+        if (GuiTab.tabRegistry.get(clazz).contains(VANILLA_INVENTORY_TAB) == false)
```
Comment (liach) [inline src/main/java/net/minecraftforge/client/player/inventory/tabs/GuiTab.java]:
```java
if (!GuiTab.tabRegistry.get(clazz).contains(VANILLA_INVENTORY_TAB))
```

File: src/main/java/net/minecraftforge/client/player/inventory/tabs/GuiTab.java [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+        }
+        RenderHelper.enableGUIStandardItemLighting();
+        GlStateManager.disableLighting();
+        GlStateManager.color(1F, 1F, 1F); // Reset color in case Items change it.
+        GlStateManager.enableBlend(); // Make sure blend is enabled else tabs show a white border.
+        guiContainer.drawTexturedModalRect(x, y, texX, texY, 28, 32);
+        x = x + 6;
+        y = y + 8 + (isFirstRow ? 1 : -1);
+        GlStateManager.enableLighting();
+        if (getIconStack() != null)
```
Comment (liach) [inline src/main/java/net/minecraftforge/client/player/inventory/tabs/GuiTab.java]: Use `!isEmpty()` check

**Review by liach — CHANGES_REQUESTED** [review]

File: src/main/java/net/minecraftforge/client/player/inventory/tabs/GuiTab.java (hunk 4 above)
Comment (AshIndigo) [reply src/main/java/net/minecraftforge/client/player/inventory/tabs/GuiTab.java]: I'll remove them if you want but any reason? Just curious

Comment (williewillus) [reply src/main/java/net/minecraftforge/client/player/inventory/tabs/GuiTab.java]: Just convention, it clutters up the code and we can give credit using the commit logs

File: src/main/java/net/minecraftforge/client/player/inventory/tabs/GuiTab.java (hunk 5 above)
Comment (AshIndigo) [reply src/main/java/net/minecraftforge/client/player/inventory/tabs/GuiTab.java]: Fixed in the latest commit

File: src/main/java/net/minecraftforge/client/player/inventory/tabs/GuiTab.java (hunk 6 above)
Comment (AshIndigo) [reply src/main/java/net/minecraftforge/client/player/inventory/tabs/GuiTab.java]: Fixed

File: src/main/java/net/minecraftforge/client/player/inventory/tabs/GuiTab.java (hunk 4 above)
Comment (AshIndigo) [reply src/main/java/net/minecraftforge/client/player/inventory/tabs/GuiTab.java]: Alright fixed

**Comment by Alpvax** [comment]
@mezz Are there any more problems with this?

File: src/main/java/net/minecraftforge/client/player/inventory/tabs/GuiTab.java [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+
+/**
+ * Class used to add a new GuiTab
+ */
+@SideOnly(Side.CLIENT)
+public class GuiTab
+{
+    private static ListMultimap<Class<? extends GuiContainer>, GuiTab> tabRegistry = ArrayListMultimap.create();
+
+    private ItemStack iconStack = null;
```
Comment (liach) [inline src/main/java/net/minecraftforge/client/player/inventory/tabs/GuiTab.java]: Set to `ItemStack.EMPTY`

**Review by liach — APPROVED** [review]

**Comment by mezz** [comment]
Doing some testing, there are some aesthetic and usability issues.

* The last tab in a row should be aligned like the rest of the tabs so the spacing is even. It looks strange being right-aligned on its own.
 * The arrows for additional tabs should go above the GUI and have page numbers like in the creative menu.
  
  
    * Without page numbers, navigation through the tabs is very confusing.
    * By adding arrows above the tabs, this means there isn't enough space for bottom tabs on small screen. That case needs to be handled properly, while still showing bottom tabs when there is enough space for them.

* Selected tabs should come 'forward' and be lit like in the creative menu.

 * Clicking through tabs sometimes gets into areas where you can't view all the tabs, which doesn't make sense at all.
  

I think this PR is an important improvement for modpacks, but there is still a bit of work to do on polish and usability before this is ready to merge.

**Comment by AbsolemJackdaw** [comment]
I must dissaprove with the last point that mezz made, it is up to the modder to add 'backward' tabs on gui's where his tabs lead to.

**Comment by Alpvax** [comment]
@ArtixAllMighty While that is true, how would that handle tabs added by multiple mods?

**Comment by AbsolemJackdaw** [comment]
if multiple mods add a tab to a chest, you can find them there. you wont find the same tabs on your own gui though. Your own gui should however be able to lead back to the chest, from which you can move to other tabs. the Chest being a HUB of many tabs, and each attached tab being a branch to other options, and hopefully back to where it came from.

It's a tree - type system. It's like navigating windows, but having to use 'go back one folder' instead of having a browser tree on the right of your screen. 

It's not the most ergonomic option there is... I know. but I think it's better then no tabs, because it would be the first step towards massively decreasing keybinds needed for opening all those modded gui's added by 150+ mods in certain modpacks, there's only ever so much keys on a keyboard

**Comment by mezz** [comment]
I recommended a flat tab structure, where the "main" gui tabs are always shown and there is no nesting of tabs.
I think the tree structure is a bit overengineered for this problem, and poses many UI challenges that haven't been addressed.
With a flat structure you can easily navigate and there is no confusion.

**Comment by AbsolemJackdaw** [comment]
the flat structure would be ideal if it was for the player's main inventory only. (which is what it was at the very begining). once you start wanting to add tabs to other gui's it becomes complicated, and a flat structure would be out of place, if i'm not wrong. but I do agree a flat structure is better for navigation

**Comment by mezz** [comment]
How does it get more complicated? I think as long as you can't add tabs to other tab guis it should be fine.

**Comment by AbsolemJackdaw** [comment]
ah you mean we're not supposed to be able to go from one tab to another to another to another ?
yeah, if you go from main to alt, and then only back, or the main tab decides what the alt's should have as tabs and an alt gui cannot have other tabs.

(this is getting complicated to understand, but i got the drift)

**Comment by AbsolemJackdaw** [comment]
nah nevermind, i reread your flat structure a couple of times and i think i got the point =)

**Comment by Alpvax** [comment]
Not sure I quite follow the flat structure. If 2 mods add tabs to the chest GUI and you click on one of them, would the other tab show up in this new GUI? Or would you need to return to the chest tab to be able to swap back to the other tab?
How would the following work with the flat structure?

- Mod1 adds a tab to the chest gui which allows you to view the player inventory (i.e. equipment).
- Mod2 adds a tab to the player inventory which allows you to view an additional "backpack"

When the player opens the chest, would all 3 tabs be displayed? Or just chest and player inventory?
When the inventory tab is selected, would all 3 show up? Or would it keep it as just chest and inventory?

In essence, would the tabs be attached to the "open GUI" instance (i.e. the tab list doesn't change until the GUI is closed) or to the individual GUIs?

EDIT: After following this through, I understand better, and I agree with @mezz, the flat, non-changing method would be better for the end user.

**Comment by AbsolemJackdaw** [comment]
what we probably need is groups, to regroup tabs with the same interests over all the gui's registered to that group.

**Comment by Alpvax** [comment]
Alternatively we could create a GuiTabList for the first GUI opened (as this PR does) but then call a setTabList method on any GUI opened from that GUI and pass in the current tab list.

You could then extend it with registered groups (e.g. a "playerinv" group which modders could add tags to, then you could add the whole playerinv group to a GUI and it would add all the tabs).

**Comment by HenryLoenwind** [comment]
lol, the original PR hat continuities for exactly this purpose. https://github.com/MinecraftForge/MinecraftForge/pull/3349#issuecomment-259992875

**Comment by AbsolemJackdaw** [comment]
yeah... thats the problem. there's 2 ways to go about this. now we need to find the one that suits 51% iof the people to let it pass ...

**Comment by Alpvax** [comment]
For the end user's point of view I think it would be better if the tabs didn't change as you navigated through the GUIs. Which would leave us with either showing all the tabs related to any of the GUIs or showing only the tabs directly linked with the first opened GUI. The 2nd option would be the easier implementation.

**Comment by HenryLoenwind** [comment]
There are a number of tricky cases here.

e.g. GUI M_A gets GUIa M_B and X_39 attached by mod X. Then mod G attached GUI G_9 to M_B and mod H attaches GUIs M_C and H_0 to G_9.

Which tabs will be shown when showing GUI M_A? Which when showing H_0? If the addition was static, one could say "all of them" for both cases. But if some of the attaching is conditional, it gets really tricky...

One solution may really be to keep a list of tabs until the user closes all GUIs. Collect and add all tabs ever show, so it may grow when going to another tab but no tabs would go away.

Another solution would be to only allow special tab-GUIs to be attached which cannot have other tabs attached to them. Then the tab list would stay static while tabbing through.

**Comment by mezz** [comment]
I think people do not immediately understand what I mean by flat tabs so I should clarify. The best example we have already for flat tabs is the vanilla creative menu. Mods can add one or more tabs there, and navigation is very easy. You click a tab and it comes to focus, and the tabs never change no matter what page you're looking at. It only gets unwieldy when there are way too many tabs, but I don't think there will be as many tabs added to GUIs using this new system as there are creative tabs.

For this system, each "main" screen (like the player inventory or the furnace GUI) can have tabs added to it. Screens that have tabs added will display a "main" tab first to get back to that main GUI, and then the additional tabs after. Clicking one of those tabs to show a tab GUI will bring it to focus, but not change the tabs or order at all.

Similar tab systems are used by JEI and Tinker's Construct.

**Comment by Alpvax** [comment]
@mezz the creative tabs are significantly simpler as (with the exception of the inventory tab) they don't change the GUI, just the items displayed, whereas these tabs will completely change the GUI being displayed. There is also only a single list of creative tabs. So somehow we need to persist a tab list instance across multiple GUIs.

What determines whether or not something is a "main" screen? Is it just the first one opened? Which "additional tabs" should be added? Just those that are linked to the main screen or should it chain through all the tab GUIs and add their tabs to the list as well?

**Comment by mezz** [comment]
The creative tabs are simpler, but we should aim to mimic that simplicity from a player point of view.

A "main" screen is one you open initially, before changing any tabs. 
Additional tabs are whatever gets added by mods to the "main" screen.

**Comment by Alpvax** [comment]
Thinking about simplifying implementation, would it not be easier to have the tabList linked to the player?

When the player opens a GUI from being in-game, the tab list is created from that first GUI. The player can then swap tabs freely without having to worry about persisting the list across multiple GUIs, and when the GUI is exited (back to in-game) the list is set to null.

**Comment by AbsolemJackdaw** [comment]
mezz is more right then any of you i.m.o

it's the player's point of view we need to aim to, and 'static' tabs around one 'main' gui is what we need. per gui.

just like. 

fancy example : baubles inventory, spacecraft and rpg inventory. 
these could all 3 register to the player inventory. this gives 4 tabs. player, baubles, spacecraft nad rpginventory tabs.
these 4 need to be registered to a list that will not change their position.
openign either 4 of these gui's will show all 4 tabs, always in the same order.

the same would go for a block that would be placed on top of a furnace to work. 
2 tabs : furnace + mod block. both have 2 tabs. if someone else would add a block with the same properties, there woudl be 3 tabs, but one of them you wouldn't be able to go to (because the mod block is missing) > note : this check needs to be implemented by the author offcourse, to prevent a crash if said block is absent

**Comment by Alpvax** [comment]
Yes, I understand that part of it, the only part I want clarification on is what happens when one mod adds a player inventory tab to the chest. Would the baubles, spacecraft and rpginventory tabs also be added to the chest?

And my comment about attaching the list to the player was just a way of making it easier to implement:

1. Player opens GUI
2. GUI provides a list of available tabs (Probably allowing a check to see if each tab is in fact valid at this time, to allow for tabs based on items in inventory, nearby blocks, capabilities etc.)
3. List of tabs is saved to the player
4. Player changes tab, list still attached to the player, but the selected index is updated to render the tab correctly (GUIs retrieve the tab list from the player in order to render it)
5. Player exits GUI and the tab list is set to null, ready for next time.

It would remove the need to pass tab instances between GUIs.

**Comment by mezz** [comment]
>what happens when one mod adds a player inventory tab to the chest. Would the baubles, spacecraft and rpginventory tabs also be added to the chest?

First this is a strange scenario, you should probably not add a player inventory tab to the chest gui.
However in that case, the player inventory tab would not be treated specially, because the chest gui is the main gui. Extra tabs would only be added for the chest. Baubles, spacecraft, and rpginventory tabs which are normally added to the player inventory would not be added to the chest tabs.

**Comment by Alpvax** [comment]
Strange but not unreasonable to think that a mod would add access to your equipment from anywhere, and I tried to pick an edge case. That solution seems good to me though.

**Comment by Alpvax** [comment]
I do also quite like the groups approach, for example an "equipment" group could be created, then things like baubles, tinkers etc. could be added to the group, and the group could then be added to another GUI from another mod.

**Comment by mezz** [comment]
@AshIndigo are you able to update this PR?

**Comment by mezz** [comment]
Please update this PR to target 1.12.
What is the status on this PR, is it complete and ready for review again?

**Comment by pWn3d1337** [comment]
It would totally like to have this in 1.12 soon before I have to make my own implementation. Please update this.

**Comment by mezz** [comment]
@AshIndigo any response?

**Comment by mezz** [comment]
Close this, it has been updated here: #4389

**Pull request closed without merging**